instrumentation/rocketmq/rocketmq-client-java/rocketmq-client-java-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/rocketmqclientjava/v5_0/MapSetter.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.rocketmqclientjava.v5_0;
+
+import io.opentelemetry.context.propagation.TextMapSetter;
+import javax.annotation.Nullable;
+import org.apache.rocketmq.client.java.message.PublishingMessageImpl;
+
+enum MapSetter implements TextMapSetter<PublishingMessageImpl> {
+  INSTANCE;
+
+  @Override
+  public void set(@Nullable PublishingMessageImpl carrier, String key, String value) {
+    if (null == carrier) {
+      return;
+    }
+    carrier.setTraceContext(value);
```

Comment:
I suspect that this won't work correctly. The problem is that user could have configured multiple context propagators and some propagators write multiple headers so this way it is not well defined what ends up set as the trace context. If you want to use `setTraceContext` for context propagation you are probably better off directly using `W3CTraceContextPropagator` and not using propagation in instrumenter api. Though if you do that then you'll lose baggage propagation. It would be much simpler if you could just add these to message properties as in  https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/b2a4cf171f51c1b5a9a1ed499bba0f0c33d9da81/instrumentation/rocketmq/rocketmq-client/rocketmq-client-4.8/library/src/main/java/io/opentelemetry/instrumentation/rocketmqclient/v4_8/MapSetter.java#L19

Reply:
OK

Reply:
`PublishingMessageImpl` here does not provide the method to extend user properties. We plan to add this feature in the recent release of `rocketmq-client-java`.

Reply:
Usually modifying the library isn't an option for us, so we have come up with a couple of strategies that could be used in such cases. For example spring integration instrumentation in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-integration-4.1/library/src/main/java/io/opentelemetry/instrumentation/spring/integration/TracingChannelInterceptor.java has the same issue. There the instrumentation makes a copy of the original message with added context propagation headers. Or you could replace the original message with a wrapper that delegates all methods to original message except `getProperties` for which it returns the modified properties.

Reply:
Thx for your opinions. I think I have achieved it by `VirtualField` rather than modifying the library.